Production bundle error when running Hilla tests

### Description of the bug

```
[INFO] --- hilla:2.5-SNAPSHOT:build-frontend (default) @ tests-spring-security-jwt ---
[INFO] Reflections took 425 ms to scan 143 urls, producing 9972 keys and 53191 values
[INFO] Scanning classes to find frontend configurations and dependencies...
[INFO] Visited 1000 classes. Took 347 ms.
[INFO] Checking if a production mode bundle build is needed
[ERROR] Error when checking if a production bundle build is needed
java.lang.NumberFormatException: 'file:../../../../ts/form' is not a valid version!
    at com.vaadin.flow.server.frontend.FrontendVersion.<init> (FrontendVersion.java:160)
    at com.vaadin.flow.server.frontend.FrontendVersion.<init> (FrontendVersion.java:128)
    at com.vaadin.flow.server.frontend.BundleValidationUtil.versionAccepted (BundleValidationUtil.java:452)
    at com.vaadin.flow.server.frontend.BundleValidationUtil.lambda$hashAndBundleModulesEqual$2 (BundleValidationUtil.java:434)
...
[INFO] Copying frontend resources from jar files ...
[INFO] Visited 143 resources. Took 38 ms.
[ERROR] Require bundle build due to missing 'config/needs-build' file.
[INFO] Running Vite ...
```



### Expected behavior

No errors

### Minimal reproducible example

https://github.com/vaadin/hilla/pull/1755

### Versions

- Vaadin / Flow version: 24.3.0.alpha
- Java version:
- OS version:
- Browser version (if applicable):
- Application Server (if applicable):
- IDE (if applicable):

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/BundleValidationUtil.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/BundleValidationUtil.java
@@ -13,6 +13,7 @@ import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 import java.util.stream.Collectors;
 
@@ -449,8 +450,30 @@ public final class BundleValidationUtil {
     }
 
     private static boolean versionAccepted(String expected, String actual) {
-        FrontendVersion expectedVersion = new FrontendVersion(expected);
-        FrontendVersion actualVersion = new FrontendVersion(actual);
+        FrontendVersion expectedVersion;
+        try {
+            expectedVersion = new FrontendVersion(expected);
+        } catch (NumberFormatException ex) {
+            expectedVersion = null;
+        }
+        FrontendVersion actualVersion;
+        try {
+            actualVersion = new FrontendVersion(actual);
+        } catch (NumberFormatException ex) {
+            actualVersion = null;
+        }
+
+        if (expectedVersion == null && actualVersion == null) {
+            return Objects.equals(expected, actual);
+        } else if (expectedVersion == null || actualVersion == null) {
+            // expected or actual version is referencing a local package
+            // while the other one is a parsable version
+            getLogger().debug(
+                    "Version '{}' cannot be parsed and compared to '{}'",
+                    expectedVersion == null ? expected : actual,
+                    expectedVersion == null ? actual : expected);
+            return false;
+        }
 
         if (expected.startsWith("~")) {
             boolean correctRange = expectedVersion

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/BundleValidationTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/BundleValidationTest.java
@@ -39,6 +39,7 @@ import com.vaadin.flow.theme.ThemeDefinition;
 import elemental.json.Json;
 import elemental.json.JsonArray;
 import elemental.json.JsonObject;
+
 import static com.vaadin.flow.server.Constants.DEV_BUNDLE_JAR_PATH;
 import static com.vaadin.flow.server.Constants.PROD_BUNDLE_JAR_PATH;
 
@@ -1792,6 +1793,91 @@ public class BundleValidationTest {
                 depScanner, finder, mode);
         Assert.assertFalse(
                 "Shouldn't re-bundle when old @vaadin/flow-frontend package is in package.json",
+                needsBuild);
+    }
+
+    @Test
+    public void localPackageInPackageJson_notChanged_noBundleRebuild()
+            throws IOException {
+        createPackageJsonStub(
+                "{\"dependencies\": {\"my-pkg\": \"file:my-pkg\"}, \"vaadin\": { \"hash\": \"aHash\"} }");
+
+        final FrontendDependenciesScanner depScanner = Mockito
+                .mock(FrontendDependenciesScanner.class);
+
+        JsonObject stats = getBasicStats();
+        stats.getObject(PACKAGE_JSON_DEPENDENCIES).put("my-pkg", "file:my-pkg");
+
+        setupFrontendUtilsMock(stats);
+
+        boolean needsBuild = BundleValidationUtil.needsBuild(options,
+                depScanner, finder, mode);
+        Assert.assertFalse(
+                "Shouldn't re-bundle when referencing local packages in package.json",
+                needsBuild);
+    }
+
+    @Test
+    public void localPackageInPackageJson_differentReference_bundleRebuild()
+            throws IOException {
+        createPackageJsonStub(
+                "{\"dependencies\": {\"my-pkg\": \"file:my-pkg\"}, \"vaadin\": { \"hash\": \"aHash\"} }");
+
+        final FrontendDependenciesScanner depScanner = Mockito
+                .mock(FrontendDependenciesScanner.class);
+
+        JsonObject stats = getBasicStats();
+        stats.getObject(PACKAGE_JSON_DEPENDENCIES).put("my-pkg",
+                "./another-folder");
+
+        setupFrontendUtilsMock(stats);
+
+        boolean needsBuild = BundleValidationUtil.needsBuild(options,
+                depScanner, finder, mode);
+        Assert.assertTrue(
+                "Should re-bundle when local packages have different values",
+                needsBuild);
+    }
+
+    @Test
+    public void localPackageInPackageJson_parsableVersionInStats_bundleRebuild()
+            throws IOException {
+        createPackageJsonStub(
+                "{\"dependencies\": {\"my-pkg\": \"file:my-pkg\"}, \"vaadin\": { \"hash\": \"aHash\"} }");
+
+        final FrontendDependenciesScanner depScanner = Mockito
+                .mock(FrontendDependenciesScanner.class);
+
+        JsonObject stats = getBasicStats();
+        stats.getObject(PACKAGE_JSON_DEPENDENCIES).put("my-pkg", "1.0.0");
+
+        setupFrontendUtilsMock(stats);
+
+        boolean needsBuild = BundleValidationUtil.needsBuild(options,
+                depScanner, finder, mode);
+        Assert.assertTrue(
+                "Should re-bundle when local package in package.json but parsable version in stats",
+                needsBuild);
+    }
+
+    @Test
+    public void localPackageInStats_parsableVersionInPackageJson_bundleRebuild()
+            throws IOException {
+        createPackageJsonStub(
+                "{\"dependencies\": {\"my-pkg\": \"1.0.0\"}, \"vaadin\": { \"hash\": \"aHash\"} }");
+
+        final FrontendDependenciesScanner depScanner = Mockito
+                .mock(FrontendDependenciesScanner.class);
+
+        JsonObject stats = getBasicStats();
+        stats.getObject(PACKAGE_JSON_DEPENDENCIES).put("my-pkg", "file:my-pkg");
+
+        setupFrontendUtilsMock(stats);
+
+        boolean needsBuild = BundleValidationUtil.needsBuild(options,
+                depScanner, finder, mode);
+        Assert.assertTrue(
+                "Should re-bundle when local package in stats but parsable version in package.json",
                 needsBuild);
     }
 

